rfcs/20240718-synthetic-benchmark/README.md

```diff
@@ -0,0 +1,279 @@
+  _bandwidth/max\_measured\_bandwidth_ for memory bound workloads, or a maximum
+  of the two for a mixed workload. For many sample sets, this metric this will
+  be quite effective.
+
+For what data should be collected, the proposal is to use psuedo-random sampling
+to generate a statistically significant distribution. We can then use simple
+metrics for analyzing the distribution for a passing state, for example by
+requiring mean performance to not regress and limiting tail performance
+regression. Because of this, we only need to benchmark a limited number of
+workloads, and the number of workloads can change based on accuracy
+requirements. Benchmarking in the range of 1,000 - 100,000 workloads should be
+sufficient for developers. In addition, since accuracy on a single workload is
+not relevant, each problem only needs to be executed a few times depending on
+caching assumptions and run to run variation requirements, allowing for
+significant benchmarking throughput relative to higher fidelity performance
+benchmarking. As such, a statistically significant distribution should be
+collectible within a few minutes.
+
+## Collecting Relevant Data For Developers
+One of the biggest issues for synthetic data is making sure it is relevant to
+current development tasks. Most development work targets specific classes of
+problems, and if sampling measures a lot of unchanged implementations, it will
+be difficult to extract useful data from run to run noise (along with the
+unchanged measurements just wasting time). To address this, we need some
+methods to limit the search space close to the optimization space. To accomplish
+this, this RFC proposes two methods be implemented:
+
+* Developer supplied filters: Developer supplied filters are expected to fit
+  into two categories, value sets or dimension restrictions. Value sets will
+  just be lists of allowed values such as the data type configuration of
+  `{s8:s8:s32,u8:u8:u32}`. Dimension restriction, on the other handle, will be
+  composed of range and modulo restrictions, for example, `{m in [1,32]:
+  m%4==0}`. In addition, as most of the development complexity arises around
+  small dimensions sizes (and large problems take longer to benchmark), indices
+  are expected to be sampled from a power law distribution.
+
+* Automated Diffing: As one of the relevant metrics is a comparison between
+  oneDNN builds, this tool needs a way to execute multiple oneDNN libraries. As
+  a consequence, we can add a testing API for comparing implementations. When
+  implementations match between oneDNN builds, we can often skip benchmarking.
+  As oneDNN primitive creation and dispatching can be quite complicated, this is
+  beneficial for developers as they cannot always predict the effect of an
+  implementation change. In addition, if functionality validation is supported,
+  this provides a much faster way to initially validate correctness by skipping
+  unchanged implementations.
```

Comment:
You're suggesting that the tool could be used to compare performance (or correctness) differences between two versions of oneDNN? How would you handle changes to the API (breaking or otherwise) between the versions? It seems to me like this feature would add unnecessary complexity.

Reply:
> How would you handle changes to the API (breaking or otherwise) between the versions?

As things currently stand, this would not be supported. The main intention here is for a fast to run regression test on a PR, not for testing between something like oneDNN releases. The problem, as I see it, is that after an optimization is made, a (most likely small) percentage of workloads are actually changed and we need to validate that optimization. If we can collect performance of 200 changed cases, we would have effective evidence for how the PR is performing in practice. This can be done in under a minute, but only if we can quickly identify what has changed. In addition, by using an automated search, we can remove accidental bias caused by developers restricting the search space to cases they believe are effected, which may not align with the reality.